Zend/zend_compile.c

```diff
@@ -7146,25 +7188,263 @@ void zend_compile_func_decl(znode *result, zend_ast *ast, bool toplevel) /* {{{
+	}
+
+	bool has_implicit_get = false;
+	bool has_implicit_set = false;
+	bool has_any_explicit_accessor = false;
+	for (uint32_t i = 0; i < accessors->children; i++) {
+		zend_ast_decl *accessor = (zend_ast_decl *) accessors->child[i];
+		zend_string *name = accessor->name;
+		zend_ast_list *param_list =
+			accessor->child[0] ? zend_ast_get_list(accessor->child[0]) : NULL;
+		zend_ast **stmt_ast_ptr = &accessor->child[2];
+		zend_ast **return_ast_ptr = &accessor->child[3];
+		zend_ast *orig_stmt_ast = *stmt_ast_ptr;
+		CG(zend_lineno) = accessor->start_lineno;
+		bool reset_return_ast = false, reset_param_type_ast = false;
+		uint32_t accessor_kind;
+
+		uint32_t accessor_visibility = accessor->flags & ZEND_ACC_PPP_MASK;
+		uint32_t prop_visibility = prop_info->flags & ZEND_ACC_PPP_MASK;
+		if (!accessor_visibility) {
+			/* Inherit the visibility of the property. */
+			accessor->flags |= prop_visibility;
+		} else {
+			if (accessor_visibility < prop_visibility) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Visibility of accessor cannot be higher than visibility of property");
+			}
+			if (accessor_visibility == prop_visibility) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Explicit accessor visibility cannot be the same as property visibility. "
+					"Omit the explicit accessor visibility");
+			}
+		}
+
+		if (ce->ce_flags & ZEND_ACC_INTERFACE) {
+			if (accessor->flags & (ZEND_ACC_PROTECTED|ZEND_ACC_PRIVATE)) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Accessor in interface cannot be protected or private");
+			}
+			if (accessor->flags & ZEND_ACC_ABSTRACT) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Accessor in interface cannot be explicitly abstract. "
+					"All interface members are implicitly abstract");
+			}
+			accessor->flags |= ZEND_ACC_ABSTRACT;
+		} else if (prop_info->flags & ZEND_ACC_ABSTRACT) {
+			if (accessor->flags & ZEND_ACC_ABSTRACT) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Accessor on abstract property cannot be explicitly abstract");
+			}
+			accessor->flags |= ZEND_ACC_ABSTRACT;
+		}
+
+		if (accessor->flags & ZEND_ACC_STATIC) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Accessor cannot be static");
+		}
+		if (prop_info->flags & ZEND_ACC_STATIC) {
+			// TODO: This restriction can be relaxed.
+			zend_error_noreturn(E_COMPILE_ERROR, "Cannot declare accessors for static property");
+		}
+		if ((accessor->flags & ZEND_ACC_FINAL) && (accessor->flags & ZEND_ACC_PRIVATE)) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Accessor cannot be both final and private");
+		}
+		if ((prop_info->flags & ZEND_ACC_FINAL) && (accessor->flags & ZEND_ACC_FINAL)) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Accessor on final property cannot be explicitly final");
+		}
+		if (accessor->flags & ZEND_ACC_ABSTRACT) {
+			if (orig_stmt_ast) {
+				zend_error_noreturn(E_COMPILE_ERROR, "Abstract accessor cannot have body");
+			}
+			if (accessor->flags & ZEND_ACC_PRIVATE) {
+				// TODO Trait exception?
```

Comment:
I don't think a trait should be able to impose a private visibility. But a trait implementing class should be able to override a `protected abstract` property visibility and decide to implement the abstract property itself with a private visibility.
I.e. the trait itself should not care about who can implement its property. It's none of his business and we should not make it seem like the trait had any say in that.

In short: if there is an abstract property, the class which implements the trait shall be able to a) define the property with any visibility or b) not implement it at all and the default visibility of the property is retained. In scenario b) `private` does not make any sense and hence should be forbidden.

Reply:
Note that traits can already define `private abstract` methods, so I think allowing this would be in line with the current behavior. https://3v4l.org/FA0Ar#v8.1.5